optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/cloner/DeepCloningUtils.java

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.domain.solution.cloner;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.ConcurrentMap;
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.optaplanner.core.api.domain.solution.cloner.DeepPlanningClone;
+import org.optaplanner.core.impl.domain.common.ReflectionHelper;
+import org.optaplanner.core.impl.domain.solution.descriptor.SolutionDescriptor;
+
+public class DeepCloningUtils {
+    private DeepCloningUtils() {
+    }
+
+    // owningClass != declaringClass
+    // owningClass refers to the actual instance class, which may be a subclass of the declaring class
+
+    /**
+     * Gets the memoized deep cloning decision from the provided
+     * memoization maps, computing them if neccessary.
+     *
+     * @param solutionDescriptor The solution descriptor for the solution class
+     * @param field The field to get the deep cloning decision of
+     * @param owningClass The class that owns the field; can be different
+     *        from the field's declaring class (ex: subclass).
+     * @param actualValueClass The class of the value that is currently assigned
+     *        to the field; can be different from the field type
+     *        (ex: for the field "List myList", the actual value
+     *        class might be ArrayList).
+     * @param fieldDeepClonedMemoization Memoization map for (field, owningClass) decisions.
+     * @param actualValueClassDeepClonedMemoization Memoization maps for actualValueClass decisions.
```

Comment:
OptaPlanner Javadoc convention: param doesn't start with an uppercase letter, therefore the whole expression is not a sentence. Therefore no full stop at the end.